piv_identity.go

```diff
@@ -0,0 +1,131 @@
+func PivIdentities() ([]PivIdentity, error) {
+	cards, err := piv.Cards()
+	if err != nil {
+		return nil, err
+	}
+	var identities []PivIdentity
+	for _, card := range cards {
+		yk, err := piv.Open(card)
+		if err != nil {
+			continue
+		}
+		cert, err := yk.Certificate(piv.SlotSignature)
+		if err != nil {
+			continue
+		}
+		if cert != nil {
+			ident := PivIdentity{card: card, yk: yk}
+			identities = append(identities, ident)
+		}
+	}
+	return identities, nil
+}
+
+// PivIdentity is an entity identity stored in a hardware key PIV applet
+type PivIdentity struct {
+	card string
+	//pin  string
+	yk *piv.YubiKey
+}
+
+var _ certstore.Identity = (*PivIdentity)(nil)
+var _ crypto.Signer = (*PivIdentity)(nil)
+
+// Certificate implements the certstore.Identity interface
+func (ident *PivIdentity) Certificate() (*x509.Certificate, error) {
+	return ident.yk.Certificate(piv.SlotSignature)
+}
+
+// CertificateChain implements the certstore.Identity interface
+func (ident *PivIdentity) CertificateChain() ([]*x509.Certificate, error) {
+	return []*x509.Certificate{}, nil
+}
+
+// Signer implements the certstore.Identity interface
+func (ident *PivIdentity) Signer() (crypto.Signer, error) {
+	return ident, nil
+}
+
+// Delete implements the certstore.Identity interface
+func (ident *PivIdentity) Delete() error {
+	panic("deleting identities on PIV applet is not supported")
+}
+
+// Close implements the certstore.Identity interface
+func (ident *PivIdentity) Close() {
+	_ = ident.yk.Close()
+}
+
+// Public implements the crypto.Signer interface
+func (ident *PivIdentity) Public() crypto.PublicKey {
+	cert, err := ident.Certificate()
+	if err != nil {
+		return nil
+	}
+
+	return cert.PublicKey
+}
+
+// Sign implements the crypto.Signer interface
+func (ident *PivIdentity) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+	private, err := ident.yk.PrivateKey(piv.SlotSignature, ident.Public(), piv.KeyAuth{
+		PINPrompt: ident.promptHardwareKeyPin,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get private key for signing")
+	}
+
+	switch private.(type) {
+	case *piv.ECDSAPrivateKey:
+		return private.(*piv.ECDSAPrivateKey).Sign(rand, digest, opts)
+	default:
+		return nil, fmt.Errorf("invalid key type")
+	}
+}
+
+func (ident *PivIdentity) promptHardwareKeyPin() (string, error) {
+	tty, err := os.OpenFile("/dev/tty", os.O_RDWR, fs.ModeCharDevice)
+	if err != nil {
+		return "", err
+	}
+	defer tty.Close()
+
+	_, err = fmt.Fprintf(tty, "enter PIN for hardware key \"%v\" (press enetr for the default PIN):\n", ident.card)
+	if err != nil {
+		return "", err
+	}
+
+	var hardwareKeyPin string
+	pin, err := terminal.ReadPassword(int(tty.Fd()))
+	if err != nil {
+		return "", err
+	}
+
+	if len(pin) == 0 {
+		hardwareKeyPin = piv.DefaultPIN
+	} else {
+		hardwareKeyPin = string(pin)
+	}
+
+	_, _ = fmt.Fprintf(tty, "Touch %v now unlock\n", ident.card)
```

Comment:
This seems specific to Security Key devices with a PIV applet. Does this text need to be changed to consider other smart-card PIV devices (like an actual SmartCard) that do not a user-presence for touch?

Or is this functionality for YubiKeys only?

Reply:
The library that I used here is intended for Yubikeys.